src/coins.cpp

```diff
@@ -503,13 +539,149 @@ void CCoinsViewCache::SetBestBlock(const uint256 &hashBlockIn) {
     hashBlock = hashBlockIn;
 }
 
+void CCoinsViewCache::AddCswNullifier(const uint256& scId,
+                         const libzendoomc::ScFieldElement &nullifier) {
+    std::pair<uint256, libzendoomc::ScFieldElement> position = std::make_pair(scId, nullifier);
+    std::pair<CCswNullifiersMap::iterator, bool> ret = cacheCswNullifiers.insert(std::make_pair(position, CCswNullifiersCacheEntry()));
+    ret.first->second.flag = CCswNullifiersCacheEntry::FRESH;
+}
+
+void CCoinsViewCache::RemoveCswNullifier(const uint256& scId,
+                         const libzendoomc::ScFieldElement &nullifier) {
+    std::pair<uint256, libzendoomc::ScFieldElement> position = std::make_pair(scId, nullifier);
+    std::pair<CCswNullifiersMap::iterator, bool> ret = cacheCswNullifiers.insert(std::make_pair(position, CCswNullifiersCacheEntry()));
+    ret.first->second.flag = CCswNullifiersCacheEntry::ERASED;
+}
+
+bool CCoinsViewCache::HaveCswNullifier(const uint256& scId,
+                         const libzendoomc::ScFieldElement &nullifier) const {
+    std::pair<uint256, libzendoomc::ScFieldElement> position = std::make_pair(scId, nullifier);
+    
+    CCswNullifiersMap::iterator it = cacheCswNullifiers.find(position);
+    if (it != cacheCswNullifiers.end())
+        return (it->second.flag = CCswNullifiersCacheEntry::ERASED);
+
+    bool tmp = base->HaveCswNullifier(scId, nullifier);
+
+    if (tmp) {
+        CCswNullifiersCacheEntry entry;
+        entry.flag = CCswNullifiersCacheEntry::DEFAULT;
+        cacheCswNullifiers.insert(std::make_pair(position, entry));
```

Comment:
the return value should be checked: if ret.second == false it means the insertion did not take place. Anyway why overwrite the entry?

Reply:
done